zsh completion: move to runtime

qmega opened the pull request:
This moves all of the logic for generating zsh completion into the completion function itself, so nothing needs to be done at compile time. This means it's no longer incompatible with cross-compilation. (IIRC, @wm4 never liked the way it worked in the first place.)

Also made some improvements to the actual functionality of the script, which are listed in the last commit message if anyone's interested. The biggest one is that it completes values for options that take `--option=help`. To facilitate this, I added a new `M_OPT_HELP` to indicate these options in the `--list-options` output, and made what I hope are acceptable changes to the output of a few options: removing some redundant column headers and changing some that printed all on one line to print one value per line.

Since the completion function itself now runs mpv and parses its output on the first tab press (after which it's cached), the completion results will not appear for as long as that takes. However, it will only do this if the argument being completed looks like an option (i.e. starts with "-"), so there is never a delay when just completing a file name. I've also put some effort into making it reasonably fast; on my machine it's consistently under 100 ms, more than half of which is mpv itself. (Having zsh compiled with PCRE support helps a lot.)

TOOLS/zsh.pl is no longer needed and is removed; a static completion script is just installed into the proper location. I removed the `--enable-zsh-comp` option as well and just install it unconditionally, since it's nothing more than copying a file. Several other static files from etc/ are installed like that, but if anyone really doesn't want it, it can be disabled by setting zshdir to empty: `./waf configure --zshdir= ...`

**Edit:** Forgot to mention, I'm no longer using `--no-config` when calling mpv, because that prevents things like profiles from being completed. It was originally added because `msg-module` in a config file would affect --list-options (see #1731), but now it appears not to.

qmega commented:
Rebased.

eintw1ck commented on options/m_option.h:
```diff
@@ -394,11 +394,14 @@ struct m_option {
 #define M_OPT_NOPROP            (1 << 6)
 
+// This options prints possible values when called as --option=help
+#define M_OPT_HELP              (1 << 7)
```
Why not just add this as `(1 << 9)` then the other flags don't need to be changed.

pavelxdd replied:
@eintw1ck to keep `M_OPT_` defines separated from `UPDATE_` defines, I suppose. The list of defines looks nicer this way.

qmega replied:
Yeah, I wanted to keep them in order and in their respective groups. In this history these have been reordered and old ones replaced sometimes, so I didn't think it'd be a problem to bump them. The values are only used internally.

eintw1ck commented on options/m_option.c:
```diff
@@ -2080,10 +2079,9 @@ static int parse_afmt(struct mp_log *log, const m_option_t *opt,
 
     if (!bstrcmp0(param, "help")) {
-        mp_info(log, "Available formats:");
+        mp_info(log, "Available formats:\n");
```
Why do there need to be newlines here?

qmega replied:
So I can avoid special cases or more complicated logic in the completion script. Most options print one value per line, so if I make these do that too then the parsing logic stays simple. I didn't see any particular reason to keep them on one line.

pavelxdd commented:
With recent changes in `m_option.c`, the `M_OPT_HELP` flag can be added to `OPT_CHOICE`, `OPT_COLOR`, `OPT_GEOMETRY` and `OPT_SIZE_BOX` options.
Although `OPT_FLAG` supports `help` parameter now as well, it doesn't need `M_OPT_HELP` flag IMO. 

There are some other options, like `hwdec` or `msg-level` that need `M_OPT_HELP` flag as well.

wm4 commented:
Why do you want to litter everything with M_OPT_HELP?

qmega commented:
@pavelxdd I only added the `M_OPT_HELP` flag to options that the completion script can actually parse values for. It would be able to do that for `OPT_CHOICE`, but it already gets values for those in a simpler way anyway. The other ones you mention don't have a set list of possible values and just print human descriptions of how to use them; the completion script would produce either no results or garbage results for those. Are you thinking that `[help]` should actually indicate anything that has a valid `=help`? That would just end up being almost everything, right? Then something different would be needed for the completion script. I don't think that would really be worth it. If help is a confusing name, it could be something else.

@wm4 I assume you were talking to Pavel, but I can't quite tell. If you were talking to me, the answer is I don't have any other way of knowing in the script that I can get values that way? Other than hardcoding the option names, which I can't imagine anyone would prefer. This is basically equivalent to `M_OPT_FILE` which was added for similar reasons.

Another option would be to add `--shell-completion=zsh` and have mpv just directly print the option definitions to be used for `_arguments`. That ought to be faster too. Would that be preferred?

pavelxdd commented:
It seems like I misunderstood something about these `M_OPT_` flags and what they are use for. Sorry guys.

wm4 commented:
My question was in general. I don't really appreciate that the script somehow tries to parse the output of `--option=help`, which doesn't return a standardized format. Thus it requires marking tons of options for something that is rather fragile and not recommended at all.

--list-options already lists choices (for choice options), and if anything you should probably use this (also it's no recommended either). If you care about options which don't list their choices there, you should probably improve that instead.

qmega commented:
It does use the listed choices for choice options, even before this change. For the others, not all of them have a set list of choices; some depend on actually probing something (e.g. `--profile` and `--audio-device`, off the top of my head). Things like `--scale` could be made to print their possible values in `--list-options`, but for some of them that wouldn't be possible.

I was also trying to strike some kind of balance between working well enough and not being too intrusive to mpv code not relating to shell completion. Though there's no standard for `=help` output, in practice they mostly follow a few pretty similar basic structures that aren't terribly likely to change, IMO.

I think the only way to avoid the fragility of parsing output is for mpv to directly produce this info in a format specifically intended for use by shell completion. I could probably get most of the help options working statically that way too, but it would require changing their own parsing code to avoid duplicating everything. Would you prefer that? Or do you have some other suggestion?

qmega commented:
FWIW, I've been using this locally since I wrote it (~2 months) and it's been noticeably better than the old one; no problems so far.

@haasn mentioned adding a computer-readable option dump. That's come up before with respect to the completion script, and IIRC the guy who added the original version decided it wasn't worth it. If we did that, and also used a runtime script like this one, the main consideration is that parsing with zsh is extremely slow, so it would have to be a pretty simple format. The easiest for this purpose would be to just print out definitions for `_arguments` one-per-line, but that wouldn't help other programmatic uses of options, if we had any.

I'm open to giving something like that a try at some point, but I never really got an answer on if that was a direction we wanted to go. I don't want to waste my time if it's going to be rejected anyway.

eintw1ck commented:
I don't think this is the correct way to do it. Other larger programs have options for generating completion data from within the application itself (though this should only be available if compiled in). (See the swift compiler).

qmega commented:
Yeah, I've been leaning that way too, if I understand you correctly. I'll probably try it eventually but I haven't gotten around to it and nobody else seems to care. Maybe something like `--dump-options={zsh,json,...}` and `--dump-option-values=<option-name>` which could be useful for a configuration gui as well.

The problem is that so many options have their own code for printing possible values, so it's kind of a pain to make those all work -- all that would have to be duplicated or made more generic. (This PR handles a lot of those because their human output is mostly pretty similar, but of course it's somewhat fragile.)

eintw1ck commented:
Can we close this as 'rejected'? I assume a new PR will have to be made anyway. @jeeb

qmega commented:
I'm still using this locally and rebasing when necessary, but if there's no chance of it being merged as is or with minor changes, it can be closed. I can make a new PR if I ever get around to trying the better way.

The first commit (M_OPT_FILE) is applicable to the old script too and should be non-controversial so that could be merged by itself.

wm4 commented on options/m_option.h:
```diff
@@ -404,10 +404,13 @@ char *format_file_size(int64_t size);
 #define M_OPT_HAVE_HELP         (1 << 7)
 
+// This options prints possible values when called as --option=help
+#define M_OPT_HELP              (1 << 8)
```
Isn't that exactly the same thing as above?

qmega replied:
Actually, it's not. `M_OPT_HAVE_HELP` looks like it's particularly for key-value lists to accept `help` in place of `help=`. It wasn't there when I first wrote this patch, or I'd have named mine differently. The options it applies to aren't parsable by this completion script so they shouldn't have the new option.

The option I'm adding here describes the *output* of `=help`, namely that it conforms to a rough format that the completion script can parse. I can rename it to something like `M_OPT_PARSABLE_HELP` and clarify the comment.

wm4 replied:
I can't say I like this. This stuff is not really mpv's concern. I think there are already too many funny flags that were added only for the zsh script, and have no purpose otherwise.

qmega replied:
That's fair. It's a little messy in the --list-options output, too. I don't like it either but the ad-hoc nature of mpv's options system doesn't really provide for anything else that I can think of. Though AFAIK the only other flag added for zsh is M_OPT_FILE -- if there are any others I don't think we're using them and they could be removed.

I did actually start a patch for a programmatically usable option dump, which probably just about everyone thinks is a better idea and could be useful outside of zsh completion. Kinda lost interest since nothing was getting merged here, so I didn't get to the hard part of working all the existing options into some common interface for providing valid values. That would have still kept M_OPT_FILE, but could have at least removed it from the --list-options output and wouldn't have added the new flag. I won't make any promises about getting back to working on that soon, though.

It sounds like this PR's approach is fundamentally not acceptable to you, so maybe you should just reject it (or just merge the first commit, which applies to the old code just as well). But if there's anything I can change that will make you happier with this, let me know. I'll keep using it myself regardless, as it's working in practice and doesn't get in the way of much from my perspective. It's been working well for me with only the occasional rebase.

wm4 replied:
> but the ad-hoc nature of mpv's options system

mpv must have one of the most sophisticated option parser systems across CLI programs. Which is actually bad, because it adds complexity.

> I did actually start a patch for a programmatically usable option dump, which probably just about everyone thinks is a better idea and could be useful outside of zsh completion. 

Sounds like it. There's already considerably metadata in option-info (not everything, but improving it would at least benefit much more users than a zsh autocomplete script).

> It sounds like this PR's approach is fundamentally not acceptable to you, so maybe you should just reject it (or just merge the first commit, which applies to the old code just as well).

I can merge the first and last commit (as listed by github), if that's OK with you.

qmega commented:
Okay, removed the middle commit. Replaced the `[parsable-help]` check in the completion script with a static list of options that's the same as what was in the old script, just so there's no regression in functionality. I was tempted to add more, but then I'd have ended up adding them all and nobody wants to maintain that. I'll probably just add the ones I actually use in my personal copy.

wm4 commented:
Can you add a commit that adjusts the CI scripts? They failed. You need to remove --enable-zsh-comp.

qmega commented:
Done.
